### ISSUE TITLE
Speed up checks for public IPv4

### ISSUE THREAD

**Issue by andrewthad** [andrewthad/haskell-ip#60]
I suspect that GHC generated crummy code with lots of branches for the `Net.IPv4.reserved` check. I think that the combination of XOR (for exact checks), CTZ (for conversion to 0 or 1), and OR (for accumulating the results) could lead to better results. I need to benchmark running this against 100 or so IP address, and then I'll actually have a baseline to work from.

**Comment by andrewthad** [comment]
GHC never ceases to amaze me. The source Haskell:

```
reserved :: IPv4 -> Bool
reserved (IPv4 w) =
  let a = getIPv4 $ fromOctets' 0 0 0 0
      b = getIPv4 $ fromOctets' 100 64 0 0
      c = getIPv4 $ fromOctets' 127 0 0 0
      d = getIPv4 $ fromOctets' 169 254 0 0
      e = getIPv4 $ fromOctets' 192 0 0 0
      f = getIPv4 $ fromOctets' 192 0 2 0
      g = getIPv4 $ fromOctets' 192 88 99 0
      h = getIPv4 $ fromOctets' 198 18 0 0
      i = getIPv4 $ fromOctets' 198 51 100 0
      j = getIPv4 $ fromOctets' 203 0 113 0
      k = getIPv4 $ fromOctets' 224 0 0 0
      l = getIPv4 $ fromOctets' 240 0 0 0
      m = getIPv4 $ fromOctets' 255 255 255 255
   in    mask8  .&. w == p24
      || mask12 .&. w == p20
      || mask16 .&. w == p16
      || mask8  .&. w == a
      || mask10 .&. w == b
      || mask8  .&. w == c
      || mask16 .&. w == d
      || mask24 .&. w == e
      || mask24 .&. w == f
      || mask24 .&. w == g
      || mask15 .&. w == h
      || mask24 .&. w == i
      || mask24 .&. w == j
      || mask4  .&. w == k
      || mask4  .&. w == l
      || mask32 .&. w == m
```

Becomes this in Core (with `dhex-word-literals`):

```
$wreserved
  = \ ww_sEMJ ->
      case and# 0xff000000## ww_sEMJ of wild_Xq8 {
        __DEFAULT ->
          case and# 0xfff00000## ww_sEMJ of {
            __DEFAULT ->
              case and# 0xffff0000## ww_sEMJ of wild2_Xqg {
                __DEFAULT ->
                  case wild_Xq8 of wild3_Xqk {
                    __DEFAULT ->
                      case and# 0xffc00000## ww_sEMJ of {
                        __DEFAULT ->
                          case wild3_Xqk of {
                            __DEFAULT ->
                              case wild2_Xqg of {
                                __DEFAULT ->
                                  case and# 0xffffff00## ww_sEMJ of wild7_XqA {
                                    __DEFAULT ->
                                      case and# 0xfffe0000## ww_sEMJ of {
                                        __DEFAULT ->
                                          case wild7_XqA of {
                                            __DEFAULT ->
                                              case and# 0xf0000000## ww_sEMJ of {
                                                __DEFAULT ->
                                                  case and# 0xffffffff## ww_sEMJ of {
                                                    __DEFAULT -> False;
                                                    0xffffffff## -> True
                                                  };
                                                0xe0000000## -> True;
                                                0xf0000000## -> True
                                              };
                                            0xc6336400## -> True;
                                            0xcb007100## -> True
                                          };
                                        0xc6120000## -> True
                                      };
                                    0xc0000000## -> True;
                                    0xc0000200## -> True;
                                    0xc0586300## -> True
                                  };
                                0xa9fe0000## -> True
                              };
                            0x7f000000## -> True
                          };
                        0x64400000## -> True
                      };
                    0x0## -> True
                  };
                0xc0a80000## -> True
              };
            0xac100000## -> True
          };
        0x0a000000## -> True
      }
```

Notice that several of the mask checks are effectively commoned up (like the `/24` masks which show up and `and#`ing with `0xff000000`. This is pretty neat, and I had not considered this possibility for optimization. Oddly, some opportunities are missed. I wonder if reordering things would help out.

**Comment by andrewthad** [comment]
If I reorder the checks as:

```
   in    mask4  .&. w == k
      || mask4  .&. w == l
      || mask8  .&. w == p24
      || mask8  .&. w == a
      || mask8  .&. w == c
      || mask10 .&. w == b
      || mask12 .&. w == p20
      || mask15 .&. w == h
      || mask16 .&. w == d
      || mask16 .&. w == p16
      || mask24 .&. w == e
      || mask24 .&. w == f
      || mask24 .&. w == g
      || mask24 .&. w == i
      || mask24 .&. w == j
      || mask32 .&. w == m
```

The Core becomes even nicer. However, performance stays the same.

**Comment by andrewthad** [comment]
Something else I've noticed is that GHC foolishly duplicates the body of `reserved` in `public`. There is no need to do this.

**Comment by andrewthad** [comment]
This was resolved by b2612866ca0af97c97004d141e7d6c27fd70f1df.
